File: aat-android/src/main/java/ch/bailu/aat/preferences/location/SolidGoToLocation.java
package ch.bailu.aat.preferences.location;

import android.content.Context;

import org.mapsforge.core.model.LatLong;

import ch.bailu.aat_lib.coordinates.CH1903Coordinates;
import ch.bailu.aat_lib.coordinates.Coordinates;
import ch.bailu.aat_lib.coordinates.OlcCoordinates;
import ch.bailu.aat_lib.coordinates.UTMCoordinates;
import ch.bailu.aat_lib.coordinates.WGS84Coordinates;
import ch.bailu.aat.map.MapViewInterface;
import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.views.preferences.SolidTextInputDialog;
import ch.bailu.aat_lib.exception.ValidationException;
import ch.bailu.aat_lib.logger.AppLog;
import ch.bailu.aat_lib.preferences.SolidString;
import ch.bailu.aat_lib.resources.Res;

public class SolidGoToLocation extends SolidString {
    private final static String KEY = "GoToLocation";
    public SolidGoToLocation(Context c) {
        super(new Storage(c), KEY);
        this.context = c;
    }


    private final Context context;


    public Context getContext() {
        return context;
    }

    private LatLong reference = null;


    @Override
    public String getLabel() {
        return Res.str().p_goto_location();
    }


    public void goToLocationFromUser(MapViewInterface map) {

        reference = map.getMapViewPosition().getCenter();

        new SolidTextInputDialog(context, this, SolidTextInputDialog.TEXT, v -> goToLocation(map, getValueAsString()));
    }

    public void goToLocation(MapViewInterface map, String s) {
        reference = map.getMapViewPosition().getCenter();

        try {
            map.setCenter(latLongFromString(s));

        } catch (Exception e) {
            AppLog.e(map.getMContext().getContext(), e);
        }

    }


    private LatLong latLongFromString(String code)
            throws  IllegalArgumentException, IllegalStateException {

        try {
            if (reference != null)
                return new OlcCoordinates(code, reference).toLatLong();

            else
                return new OlcCoordinates(code).toLatLong();


        } catch (Exception eOLC) {
            try {
                return new CH1903Coordinates(code).toLatLong();

            } catch(Exception eCH1903) {
                try {
                    return new WGS84Coordinates(code).toLatLong();

                } catch (Exception eWGS) {
                    try {
                        return new UTMCoordinates(code).toLatLong();
                    } catch(Exception eUTM) {
                        throw Coordinates.getCodeNotValidException(code);
                    }
                }
            }
        }
    }


    @Override
    public void setValueFromString(String s) throws ValidationException {
        s = s.trim();


        if (! validate(s)) {
            throw new ValidationException(Res.str().p_goto_location_hint());
        } else {
            setValue(s);
        }
    }


    @Override
    public boolean validate(String s) {
        try  {
            return latLongFromString(s) != null;
        } catch (Exception e) {
            return false;
        }
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/preferences/location/SolidGoToLocation.kt
package ch.bailu.aat.preferences.location

import android.content.Context
import ch.bailu.aat.preferences.Storage
import ch.bailu.aat.views.preferences.dialog.SolidTextInputDialog
import ch.bailu.aat_lib.coordinates.LocationParser
import ch.bailu.aat_lib.exception.ValidationException
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.map.MapViewInterface
import ch.bailu.aat_lib.preferences.SolidString
import ch.bailu.aat_lib.resources.Res
import org.mapsforge.core.model.LatLong

class SolidGoToLocation(val context: Context) : SolidString(
    Storage(context), KEY)
{
    companion object {
        private const val KEY = "GoToLocation"
    }

    private var reference: LatLong? = null

    override fun getLabel(): String {
        return Res.str().p_goto_location()
    }

    fun goToLocationFromUser(map: MapViewInterface) {
        reference = map.getMapViewPosition().center
        SolidTextInputDialog(context,this, SolidTextInputDialog.TEXT) {
            goToLocation(map, getValueAsString())
        }
    }

    fun goToLocation(map: MapViewInterface, s: String) {
        reference = map.getMapViewPosition().center
        try {
            map.setCenter(LocationParser.latLongFromString(s))
        } catch (e: Exception) {
            AppLog.e(this, e)
        }
    }

    @Throws(ValidationException::class)
    override fun setValueFromString(string: String) {
        if (!validate(string)) {
            throw ValidationException(Res.str().p_goto_location_hint())
        } else {
            setValue(string)
        }
    }
}
